llmfoundry/models/utils/meta_init_context.py

```diff
@@ -80,21 +80,27 @@ def init_on_device(device: torch.device, include_buffers: bool = False):
     if include_buffers:
         old_register_buffer = nn.Module.register_buffer
 
-    def register_empty_parameter(module: torch.nn.Module, name: str,
+    def register_empty_parameter(self: torch.nn.Module, name: str,
                                  param: Optional[torch.nn.Parameter]):
-        old_register_parameter(module, name, param)
+        old_register_parameter(self, name, param)
         if param is not None:
-            param_cls = type(module._parameters[name])
-            kwargs = module._parameters[name].__dict__  # type: ignore
-            module._parameters[name] = param_cls(
-                module._parameters[name].to(device), **kwargs)  # type: ignore
-
-    def register_empty_buffer(module: torch.nn.Module, name: str,
-                              buffer: Optional[torch.Tensor]):
-        old_register_buffer(module, name, buffer)
-        if buffer is not None:
-            module._buffers[name] = module._buffers[name].to(  # type: ignore
-                device)
+            parameter = self._parameters[name]
+            assert parameter is not None
+
+            param_cls = type(parameter)
+            kwargs = parameter.__dict__
+
+            self._parameters[name] = param_cls(parameter.to(device), **kwargs)
+
+    def register_empty_buffer(self: torch.nn.Module,
+                              name: str,
+                              tensor: Optional[torch.Tensor],
+                              persistent: bool = True):
+        old_register_buffer(self, name, tensor, persistent=persistent)
```

Comment:
New functionality here is to pass `persistent` argument to the old register buffer method. Otherwise, the patching below would be incorrect.

Reply:
I am hesitant to change this code because I believe this file is just fully copied from accelerate and im not sure if they have code making some unexpected assumptions about this code: https://github.com/huggingface/accelerate/blob/4a1d5b1fb651c1c408c39f9ddaab37c54cd878ba/src/accelerate/big_modeling.py#L85-L158, however the change looks fine. wdyt?

Reply:
Looking at the link you provided, looks like they do pass `persistent` over, as of four months ago:
```
  def register_empty_buffer(module, name, buffer, persistent=True):
 ```
 
So maybe it's OK? Defer to you all, I do not know this codebase well enough.

Reply:
good point haha, ok should be fine. thanks!